packages/block-library/src/query-pagination/index.php

```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Server-side rendering of the `core/query-pagination` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Renders the `core/query-pagination` block on the server.
+ *
+ * @param array    $attributes Block attributes.
+ * @param string   $content    Block default content.
+ * @param WP_Block $block      Block instance.
+ *
+ * @return string Returns the pagination for the query.
+ */
+function render_block_core_query_pagination( $attributes, $content, $block ) {
+	$page_key = 'query-' . $block->context['id'] . '-page';
+	$page     = empty( $_GET[ $page_key ] ) ? 1 : $_GET[ $page_key ];
+
+	$content = '';
+	if ( 1 !== $page ) {
+		$content .= sprintf(
+			'<div><a href="%s">Previous</a></div>',
+			add_query_arg( $page_key, '2' === $page ? false : $page - 1 )
+		);
+	}
+	if ( $page < $block->context['query']['pages'] ) {
+		$content .= sprintf(
+			'<div><a href="%s">Next</a></div>',
+			add_query_arg( $page_key, $page + 1 )
+		);
+	}
+	return $content;
```

Comment:
Is this block meant to handle post navigation when on `single` contexts as well?

If so, it'd make sense to replicate or borrow some of https://developer.wordpress.org/reference/functions/get_the_post_navigation/

Otherwise: https://developer.wordpress.org/reference/functions/get_posts_nav_link/

Reply:
I think these queries/paginations need to be independent of their surrounding template's 'default' context, since they could otherwise 'break out' of the template -- e.g. we could have a query in a post that shows a category archive, in which case the prev/next navigation needs to go through that category's pages, rather than the adjacent posts.

(I hope I understood your question correctly :sweat_smile: )

We should however probably make sure to use the 'proper' navigation for the template's default context. So if we're editing a category archive template, and we use an (implied or explicit) query at the top level, we should make sure that the next/prev buttons use the 'correct' links to the next/prev page in the archive.

Reply:
>If so, it'd make sense to replicate or borrow some of

Yeah, we can't use the functions directly because we can't easily manipulate the generated links, but we can borrow.

>I think these queries/paginations need to be independent of their surrounding template's 'default' context

The Query block can set its own query or inherit the global one.